In [1]:
def gradient_descent(X, y, learning_rate=0.01, iterations=1000):

    m = len(y)  # Number of samples
    n = X.shape[1]  # Number of features
    W = np.zeros(n)  # Initialize weights to zero
    costs = []

    for _ in range(iterations):
        # Calculate predictions
        Y_pred = X @ W

        # Calculate errors
        errors = Y_pred - y

        # Calculate the cost (Mean Squared Error)
        cost = (1 / (2 * m)) * np.sum(errors ** 2)
        costs.append(cost)


        # Calculate gradients
        gradients = (1 / m) * (X.T @ errors)

        # Update weights
        W = W - learning_rate * gradients

    return W, costs

In [2]:
# define rsme

import numpy as np

def rmse(y_true, y_predicted):
  mse = np.mean((y_true - y_predicted)**2)
  rmse = np.sqrt(mse)
  return rmse

In [3]:
# define data

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

try:
    df = pd.read_csv('/content/drive/MyDrive/Dataset/student.csv')
except FileNotFoundError:
    print("Error: 'student.csv' not found. Please provide the correct file path.")
    exit() # or handle the error in a different way


In [6]:
# define x

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

def gradient_descent(X, y, learning_rate=0.01, iterations=1000):
    m = len(y)  # Number of samples
    n = X.shape[1]  # Number of features
    W = np.zeros(n)  # Initialize weights to zero
    costs = []

    for _ in range(iterations):
        # Calculate predictions
        Y_pred = X @ W

        # Calculate errors
        errors = Y_pred - y

        # Calculate the cost (Mean Squared Error)
        cost = (1 / (2 * m)) * np.sum(errors ** 2)
        costs.append(cost)


        # Calculate gradients
        gradients = (1 / m) * (X.T @ errors)

        # Update weights
        W = W - learning_rate * gradients

    return W, costs
# define rsme


def rmse(y_true, y_predicted):
  mse = np.mean((y_true - y_predicted)**2)
  rmse = np.sqrt(mse)
  return rmse
# define data


try:
    df = pd.read_csv('/content/drive/MyDrive/Dataset/student.csv')
except FileNotFoundError:
    print("Error: 'student.csv' not found. Please provide the correct file path.")
    exit() # or handle the error in a different way

# Define x (features)
x = df[['Math', 'Reading']]
x = np.array(x)
# Add bias term to x
X = np.concatenate((np.ones((x.shape[0], 1)), x), axis=1)

y = df['Writing']
y = np.array(y)

In [8]:
y = df['Writing']
y = np.array(y)

In [11]:
# define r2

def r2(y_true, y_predicted):
    # Calculate the mean of the true values
    mean_y_true = np.mean(y_true)

    # Calculate the total sum of squares (TSS)
    TSS = np.sum((y_true - mean_y_true)**2)

    # Calculate the residual sum of squares (RSS)
    RSS = np.sum((y_true - y_predicted)**2)

    # Calculate R-squared
    r2 = 1 - (RSS / TSS)

    return r2

In [14]:
# define y

y = df['Writing']
y = np.array(y)

In [15]:
# Main Function
def main():
  # Step 1: Load the dataset
  data = pd.read_csv('/content/drive/MyDrive/Dataset/student.csv')
  # Step 2: Split the data into features (X) and target (Y)
  X = data[['Math', 'Reading']].values # Features: Math and Reading marks
  Y = data['Writing'].values # Target: Writing marks
  # Step 3: Split the data into training and test sets (80% train, 20% test)
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)
  # Step 4: Initialize weights (W) to zeros, learning rate and number of iterations
  W = np.zeros(X_train.shape[1]) # Initialize weights
  alpha = 0.00001 # Learning rate
  iterations = 1000 # Number of iterations for gradient descent
  # Step 5: Perform Gradient Descent
  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)
  # Step 6: Make predictions on the test set
  Y_pred = np.dot(X_test, W_optimal)
  # Step 7: Evaluate the model using RMSE and R-Squared
  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)
  # Step 8: Output the results
  print("Final Weights:", W_optimal)
  print("Cost History (First 10 iterations):", cost_history[:10])
  print("RMSE on Test Set:", model_rmse)
  print("R-Squared on Test Set:", model_r2)
# Execute the main function
if __name__ == "__main__":
  main()

Final Weights: [0.34811659 0.64614558]
Cost History (First 10 iterations): [2471.69875, 2013.165570783755, 1640.286832599692, 1337.0619994901588, 1090.479489285058, 889.9583270083235, 726.8940993009545, 594.2897260808594, 486.4552052951634, 398.7634463599482]
RMSE on Test Set: 5.2798239764188635
R-Squared on Test Set: 0.8886354462786421


Present your finding:
1. Did your Model Overfitt, Underfitts, or performance is acceptable.
2. Experiment with different value of learning rate, making it higher and lower, observe the result.

Overfitting: If the R² on the test set were much lower than on the training set, it would indicate overfitting.

Underfitting: If the R² were significantly below 0.7, the model might underfit.

Since the test set R² is high (0.89) and the RMSE is reasonable, the model generalizes well without showing symptoms of overfitting or underfitting.
The model's performance is acceptable and reflects a good balance between bias and variance